File: ServerSide/BallPassSchedulePattern/Players_QrcodeReaderTypes/Decode_b_ReplacePatternPlayerDir/ReplacePatternPlayer.py
import os, sys
import numpy as np
import copy
sys.path.append("../..")
from Players_CommonPlayers.SuperPlayerDir.SuperPlayer import SuperPlayer

class ReplacePatternPlayer(SuperPlayer):
    def __init__(self):
        super().__init__()
        self.my_name = None  # 初期化は必ずNone
        self.replaced_matrix = None
        self.png_file_path = None # 写真のパス
        self.binary_matrix_2Dlist = None # コード部分25＊25
    
    def return_my_name(self):
        return "ReplacePatternPlayer"
    
    def replace_patterns(self, matrix):
        """
        QRコードの位置検出パターン、タイミングパターン、ダークモジュールを-11に置き換える。
        25x25のQRコードのマトリックスを前提として処理する。
        """
        # numpy配列に変換して作業する
        matrix_np = np.array(matrix)
        
        # 位置検出パターン（左上、右上、左下の位置）
        matrix_np[0:7, 0:7] = -11  # 7×7の領域(と、内側の方1分)を-11で塗りつぶす
        matrix_np[0:7, -7:] = -11  
        matrix_np[-7:, 0:7] = -11  

        # タイミングパターン（縦横方向の特定の行列）
        matrix_np[6, 8:17] = -11  # 横方向
        matrix_np[8:17, 6] = -11  # 縦方向

        # ダークモジュール（特定の座標）
        matrix_np[8, 13] = -11

        # 完成した置き換え後のマトリックスを戻り値とする
        return matrix_np
    
    def main(self):
        """
        抽出後のmatrixを取得し、位置検出パターン、タイミングパターン、ダークモジュールを
        -11に置き換える。
        """
        """ 入力 """
        woP = self.one_time_world_instance.patternDetectionPlayer
        self.png_file_path = woP.png_file_path # 写真のパス
        self.binary_matrix_2Dlist = woP.binary_matrix_2Dlist # コード部分25＊25
        
        """ 初期化 """
        # FormatInfoCatcherPlayerで取得したmatrixを取得
        matrix_2Dlist = copy.deepcopy(self.one_time_world_instance.formatInfoCatcherPlayer.matrix)
        
        """ matrixの特定部分を-11に置き換え """
        self.replaced_matrix = self.replace_patterns(matrix_2Dlist)
        
        # 自身を更新
        self.one_time_world_instance.replacePatternPlayer = self
        
        """ 出力 """
        self.replaced_matrix # 置き換え後のmatrix
        self.png_file_path # 写真のパス
        self.binary_matrix_2Dlist # 置き換え前のmatrix

        return "Completed"

File: InstanceSide/BallPassSchedulePattern/Players_QrcodeReaderTypes/Decode_b_ReplacePatternPlayerDir/ReplacePatternPlayer.cs
using UdonSharp;
using UnityEngine;

public class ReplacePatternPlayer : SuperPlayer
{
    public string PngFilePath; // 写真のパス
    public int[,] BinaryMatrix2DArray; // 置き換え前のQRコード部分(25x25の2次元配列)
    public int[,] ReplacedMatrix; // 置き換え後のQRコードマトリックス(25x25の2次元配列)

    public PatternDetectionPlayer PatternDetectionPlayer; // Unityエディタでアタッチ
    public FormatInfoCatcherPlayer FormatInfoCatcherPlayer; // Unityエディタでアタッチ
    public RinaNumpy rinaNumpy; // RinaNumpyをUnityエディタでアタッチ

    public override string ReturnMyName()
    {
        return "ReplacePatternPlayer";
    }

    private int[,] ReplacePatterns(int[,] matrix)
    {
        // matrixの大きさチェック
        if (matrix.GetLength(0) != 25 || matrix.GetLength(1) != 25)
        {
            Debug.LogError("Matrix must be 25x25.");
            return new int[0, 0];
        }

        // 入力マトリックスをコピー
        int[,] result = new int[25, 25];
        rinaNumpy.CopyIntArray(matrix, result, 25 * 25); // RinaNumpyで配列をコピー

        // 位置検出パターン
        for (int i = 0; i < 7; i++)
        {
            for (int j = 0; j < 7; j++)
            {
                result[i, j] = -11; // 左上
                result[i, 25 - 7 + j] = -11; // 右上
                result[25 - 7 + i, j] = -11; // 左下
            }
        }

        // タイミングパターン
        for (int i = 8; i < 17; i++)
        {
            result[6, i] = -11; // 横方向
            result[i, 6] = -11; // 縦方向
        }

        // ダークモジュール
        result[8, 13] = -11;

        return result;
    }

    public override string ExecuteMain()
    {
        // PatternDetectionPlayerのデータを取得
        if (PatternDetectionPlayer != null)
        {
            PngFilePath = PatternDetectionPlayer.PngFilePath;
            BinaryMatrix2DArray = PatternDetectionPlayer.BinaryMatrix2DArray;
        }
        else
        {
            Debug.LogError("PatternDetectionPlayer is not attached.");
            return "Error";
        }

        // FormatInfoCatcherPlayerのデータを取得
        int[,] matrix = null;
        if (FormatInfoCatcherPlayer != null)
        {
            matrix = FormatInfoCatcherPlayer.Matrix;
        }
        else
        {
            Debug.LogError("FormatInfoCatcherPlayer is not attached.");
            return "Error";
        }

        // QRコードのパターンを置き換える
        ReplacedMatrix = ReplacePatterns(matrix);

        // 自身を更新
        PatternDetectionPlayer.ReplacePatternPlayer = this;

        return "Completed";
    }
}
